include/flux/op/cartesian_product.hpp

```diff
@@ -88,24 +91,37 @@ struct cartesian_product_traits_base {
 
     template <std::size_t I, typename Self>
     static constexpr auto ra_inc_impl(Self& self, cursor_type<Self>& cur, distance_t offset)
-    -> cursor_type<Self>&
+        -> cursor_type<Self>&
     {
-        auto& base = std::get<I>(self.bases_);
-
-        auto this_sz = flux::size(base);
-        auto this_offset = offset  % this_sz;
-        auto next_offset = offset/this_sz;
+        if (offset == 0)
+            return cur;
 
-        // Adjust this cursor by the corrected offset
-        flux::inc(base, std::get<I>(cur), this_offset);
+        auto& base = std::get<I>(self.bases_);
+        const auto this_index = flux::distance(base, flux::first(base), std::get<I>(cur));
+        auto new_index = num::checked_add(this_index, offset);
+        auto this_size = flux::size(base);
+
+        // If the new index overflows the maximum or underflows zero, calculate the carryover and fix it.
+        if (new_index < 0 || new_index >= this_size) {
+            offset = num::checked_div(new_index, this_size);
+            new_index = num::checked_mod(new_index, this_size);
+
+            // Correct for negative index which may happen when underflowing.
+            if (new_index < 0) {
+                new_index = num::checked_add(new_index, this_size);
```

Comment:
Massive pedantry: `this_size` must be positive (or zero), so if `new_index` is negative then this addition can never overflow and we don't technically need to do a `checked_add()`. But honestly it's probably not worth fixing.

Reply:
I'd rather err on the side of caution and leave the check in. There are other places in the function where the checked numerics are important. I don't want future readers or modifiers of the function to have to discern which uses are significant and which are not.